src/wp-includes/canonical.php

```diff
@@ -742,6 +756,16 @@ function lowercase_octets( $matches ) {
 		$requested_url = preg_replace_callback( '|%[a-fA-F0-9][a-fA-F0-9]|', 'lowercase_octets', $requested_url );
 	}
 
+	if (
+		$redirect_obj &&
+		is_a( $redirect_obj, 'WP_Post' ) &&
+		get_post_status_object( $redirect_obj->post_status )->public === false &&
```

Comment:
Use `get_post_status()` to avoid breaking `inherit`.

Add tests for `inherit` from `public` and `private`.